### PR TITLE
Add extra verification when verifying tags

### PR DESCRIPTION
**Pull request by neilnaveen**
- Adds an extra check that verifies that the target hash of the tag object is the same as its corresponding RSL entry's target hash.

### CODE REVIEW

File: internal/policy/verify.go [hunk 1]
```diff
@@ -410,6 +410,23 @@ func VerifyTag(ctx context.Context, repo *git.Repository, ids []string) map[stri
 			continue
 		}
 
+		tagRef, err := repo.Tag(strings.Trim(absPath, "refs/tags"))
```
Comment (adityasaky) [inline internal/policy/verify.go:413]: This could just use `repo.Reference` and avoid having to trim the prefix.

Comment (adityasaky) [reply internal/policy/verify.go:413]: Also, this should maybe be a part of `verifyTagEntry` where we also load the tag object.

File: internal/policy/verify.go [hunk 2]
```diff
@@ -410,6 +410,23 @@ func VerifyTag(ctx context.Context, repo *git.Repository, ids []string) map[stri
...
+		if err != nil {
+			status[id] = err.Error()
+			continue
+		}
+
+		tagObj, err := gitinterface.GetTag(repo, tagRef.Hash())
+
```
Comment (adityasaky) [inline internal/policy/verify.go:420]:
```suggestion
```

File: internal/policy/verify.go [hunk 3]
```diff
@@ -410,6 +410,23 @@ func VerifyTag(ctx context.Context, repo *git.Repository, ids []string) map[stri
...
+		if err != nil {
+			status[id] = err.Error()
+			continue
+		}
+
+		if entry.TargetID != tagObj.Target {
```
Comment (adityasaky) [inline internal/policy/verify.go:426]: is this right? the entry may point to the tag obj, not the target of the tag obj

File: internal/policy/verify.go (hunk 1 above)
Comment (adityasaky) [reply internal/policy/verify.go:413]: Also, perhaps worth double checking that the entry's ref matches tagObj.Name.

**Review by adityasaky — CHANGES_REQUESTED** [review]
Thanks for this patch! Some comments...

File: internal/policy/verify.go (hunk 1 above)
Comment (neilnaveen) [reply internal/policy/verify.go]: I have moved it to verifyTagEntry, and I have added a check to double check that the entry's ref matches the tagObj.Name.

File: internal/policy/verify.go (hunk 2 above)
Comment (neilnaveen) [reply internal/policy/verify.go]: Fixed

File: internal/policy/verify.go (hunk 3 above)
Comment (neilnaveen) [reply internal/policy/verify.go]: I have changed this to check if it is either pointing to the entry.TargetID, or the tagObj itself

File: internal/policy/verify.go (hunk 1 above)
Comment (neilnaveen) [reply internal/policy/verify.go]: repo.Reference still outputs in the format refs/tags/tagname, so trim would still have to be done.

File: internal/policy/verify.go [hunk 4]
```diff
@@ -682,6 +682,14 @@ func verifyTagEntry(ctx context.Context, repo *git.Repository, policy *State, en
 		return err
 	}
 
+	if entry.TargetID != tagObj.Target && entry.TargetID != tagObj.Hash {
+		return fmt.Errorf("verifying RSL entry failed, %s", ErrUnauthorizedSignature.Error())
+	}
+
+	if entry.RefName != "refs/tags/"+tagObj.Name {
+		return fmt.Errorf("verifying RSL entry failed, tag name has been changed")
```
Comment (adityasaky) [inline internal/policy/verify.go:690]: it may have changed or it may not have, it may actually be the case that refs/tags/<x> points to tag <y>. Changed is more indicative that the same tag name has a different target. I'd rephrase this to just say they don't match.

FWIW, I think Git may have some default protections for this submitted as part of the original omitting commits work.

**Review by adityasaky — CHANGES_REQUESTED** [review]
This largely looks good to me, just a couple quick comments.

I also wonder if we want to update the rsl record flow for tags to perform the same checks.

**Review by adityasaky — DISMISSED** [review]
Looks good! Can you add some tests that hit these checks?

File: internal/policy/verify.go [hunk 5]
```diff
@@ -682,6 +682,14 @@ func verifyTagEntry(ctx context.Context, repo *git.Repository, policy *State, en
 		return err
 	}
 
+	if entry.TargetID != tagObj.Hash && entry.TargetID != tagObj.Target {
+		return fmt.Errorf("verifying RSL entry failed, %s", ErrUnauthorizedSignature.Error())
+	}
+
+	if entry.RefName != plumbing.NewTagReferenceName(tagObj.Name).String() {
+		return fmt.Errorf(plumbing.NewTagReferenceName(entry.RefName).String())
```
Comment (adityasaky) [inline internal/policy/verify.go:690]: error needs to be updated? Perhaps that it doesn't match. Also, you can just use entry.RefName afaict.

Comment (neilnaveen) [reply internal/policy/verify.go]: Fixed

**Comment by neilnaveen** [comment]
I have added additional test, getting coverage for 

```
if entry.RefName != plumbing.NewTagReferenceName(tagObj.Name).String() {
 	return fmt.Errorf("verifying RSL entry failed, tag name does not match expected tag name from RSL")
 }
```

I am not sure how to test the other clause, how would I go about this? For this to be done, I would need to record two RSL entries, which would error out, since tags cannot have more than one.

**Comment by adityasaky** [comment]
this is with the standard rsl.New<>.Commit()? IIRC if you craft the entries directly, you should be fine, can you check?

**Comment by neilnaveen** [comment]
That was not actualy the problem, the problem was that I was not checking if our entry tag from the ref's hash is not equal to the tag got fromt the entry's targetHash's hash.

File: internal/policy/verify.go [hunk 6]
```diff
@@ -682,6 +682,15 @@ func verifyTagEntry(ctx context.Context, repo *git.Repository, policy *State, en
 		return err
 	}
 
+	entryTagRef, err := repo.Reference(plumbing.ReferenceName(entry.RefName), false)
```
Comment (adityasaky) [inline internal/policy/verify.go:685]: If you set this to `true`, does it resolve to the target?

File: internal/policy/verify_test.go [hunk 7]
```diff
@@ -744,39 +744,65 @@ func TestVerifyCommit(t *testing.T) {
...
+		entryID = common.CreateTestRSLReferenceEntryCommit(t, repo, entry, gpgKeyBytes)
+		entry.ID = entryID
+
+		common.CreateTestSignedTag(t, repo, tagName, commitIDs[len(commitIDs)-2], gpgKeyBytes)
+
+		// Use tag ID
+		expectedStatus := map[string]string{tagID.String(): fmt.Sprintf("verifying RSL entry failed, %s", ErrUnauthorizedSignature.Error())}
+		status := VerifyTag(context.Background(), repo, []string{tagID.String()})
+		assert.Equal(t, expectedStatus, status)
+	})
```
Comment (adityasaky) [inline internal/policy/verify_test.go:805]: This doesn't include a test for duplicate checks in the verification flow, right?

File: internal/policy/verify.go (hunk 6 above)
Comment (neilnaveen) [reply internal/policy/verify.go]: Resolving it will either keep it the same if it is the targets hash, or if it is the tag's hash, it will change it into its target, so at the end it is just better to resolve it. Fixed.

File: internal/policy/verify_test.go (hunk 7 above)
Comment (neilnaveen) [reply internal/policy/verify_test.go:805]: Sorry, could you clarify what you mean?

File: internal/policy/verify.go [hunk 8]
```diff
@@ -682,6 +682,15 @@ func verifyTagEntry(ctx context.Context, repo *git.Repository, policy *State, en
 		return err
 	}
 
+	entryTagRef, err := repo.Reference(plumbing.ReferenceName(entry.RefName), true)
+	if err != nil {
+		return err
+	}
+
+	if entry.TargetID != entryTagRef.Hash() && entry.TargetID != tagObj.Target {
+		return fmt.Errorf("verifying RSL entry failed, %s", ErrUnauthorizedSignature.Error())
```
Comment (adityasaky) [inline internal/policy/verify.go:691]: I'd not use ErrUnauthorizedSignature here, let's use an error that reflects the mismatch in metadata?

File: internal/policy/verify_test.go (hunk 7 above)
Comment (adityasaky) [reply internal/policy/verify_test.go:805]: recording tag v1 in the RSL and then recording the exact same tag again in the RSL. The RSL should enforce immutability of tags it's aware of.

**Comment by adityasaky** [comment]
Outstanding minor nits aside, we should be able to merge this soon. Sorry for the delay!

File: internal/policy/verify.go (hunk 8 above)
Comment (neilnaveen) [reply internal/policy/verify.go]: Fixed

File: internal/policy/verify_test.go (hunk 7 above)
Comment (neilnaveen) [reply internal/policy/verify_test.go:805]: I have added this test

File: internal/policy/verify.go [hunk 9]
```diff
@@ -682,6 +682,15 @@ func verifyTagEntry(ctx context.Context, repo *git.Repository, policy *State, en
 		return err
 	}
 
+	entryTagRef, err := repo.Reference(plumbing.ReferenceName(entry.RefName), true)
+	if err != nil {
+		return err
+	}
+
+	if entry.TargetID != entryTagRef.Hash() && entry.TargetID != tagObj.Target {
+		return fmt.Errorf("verifying RSL entry failed, gittuf metadata does not match tag data")
```
Comment (adityasaky) [inline internal/policy/verify.go:691]:
```suggestion
		return fmt.Errorf("verifying RSL entry failed, tag reference set to unexpected target")
```

File: internal/policy/verify_test.go [hunk 10]
```diff
@@ -744,39 +744,89 @@ func TestVerifyCommit(t *testing.T) {
...
+		entry.ID = entryID
+
+		common.CreateTestSignedTag(t, repo, tagName, commitIDs[len(commitIDs)-2], gpgKeyBytes)
+
+		// Use tag ID
+		expectedStatus := map[string]string{tagID.String(): "verifying RSL entry failed, gittuf metadata does not match tag data"}
+		status := VerifyTag(context.Background(), repo, []string{tagID.String()})
+		assert.Equal(t, expectedStatus, status)
+	})
+	t.Run("tag verification with multiple RSL entries", func(t *testing.T) {
```
Comment (adityasaky) [inline internal/policy/verify_test.go:806]:
```suggestion

	t.Run("tag verification with multiple RSL entries", func(t *testing.T) {
```

**Review by adityasaky — APPROVED** [review]
some nits, but LGTM!

**Comment by adityasaky** [comment]
Could you address the nits + clean up commits? Thanks!

File: internal/policy/verify_test.go (hunk 10 above)
Comment (neilnaveen) [reply internal/policy/verify_test.go:807]: Fixed

File: internal/policy/verify.go (hunk 9 above)
Comment (neilnaveen) [reply internal/policy/verify.go]: Fixed

**Comment by neilnaveen** [comment]
I have cleaned up the commit history and fixed the nits

**Review by adityasaky — APPROVED** [review]
LGTM, thanks!
